site/static/scripts/ui/domInteraction.js

```diff
@@ -0,0 +1,356 @@
+        elem.className = "device_row";
+        elem.id = name;
+
+        // Creates button to delete server
+        var btn = document.createElement("BUTTON");
+        btn.innerText="x";
+        btn.onclick = submitDeleteServer;
+        elem.appendChild(btn);
+
+        // Function for click on device row
+        elem.onclick = function(){
+            setAllChildrenToClass(serversListElem, "device_row");
+            this.className = "device_row active";
+
+            var server = servers[this.id];
+            
+            SELECTED_SERVER = server;
+            SELECTED_DEVICE = null;
+
+            if (updateDevices){
+                updateDeviceList();
+                updateAddDeviceModule();
+            }
+        };
+        elem.appendChild(document.createTextNode(name));
+        serversListElem.appendChild(elem);
+    }
+
+    if (SELECTED_SERVER == null){
+        if (serversListElem.firstChild != null){
+            serversListElem.firstChild.click();
+        }
+    } else {
+        document.getElementById(SELECTED_SERVER).click();
+    }
+}
+
+/**
+ * Generates the html to view all of the devices in data.
+ * @param {} data
+ */
+function populateDevices(data){
+    var namesListElem = document.getElementById("deviceNamesList");
+    removeChildren(namesListElem);
+
+    removeChildren(document.getElementById("activatorsList"));
+
+    data.forEach(function(device) {
+        var elem = document.createElement("li");
+        elem.className = "device_row";
+        elem.id = device.deviceId;
+        var btn = document.createElement("BUTTON");
+        btn.innerText="x";
+        btn.onclick = submitDeleteDevice;
+        elem.appendChild(btn);
+
+        elem.onclick = function() {
+            setAllChildrenToClass(namesListElem, "device_row");
+            elem.className = "device_row active";
+
+            var dev = getDeviceById(elem.id, data);
+            SELECTED_DEVICE = dev.deviceId;
+
+            populateDeviceActivators(dev.name, dev.deviceId, dev.activators);
+        };
+
+        elem.appendChild(document.createTextNode(device.name));
+
+        var renameButton = document.createElement("BUTTON");
+        renameButton.innerText = "Rename";
+        renameButton.onclick = submitRenameDevice;
+        elem.appendChild(renameButton);
+
+
+        namesListElem.appendChild(elem);
+    });
+
+    if (SELECTED_DEVICE == null){
+        if (namesListElem.firstChild != null){
+            namesListElem.firstChild.click();
+        }
+    } else {
+        document.getElementById(SELECTED_DEVICE).click();
+    }
+}
+
+/**
+ * Generates the html to view a device with a particular deviceName and for its
+ * associated activators. deviceId is passed so that we may interact with the
+ * activators and know their device. This will be refactored.
+ * @param {} deviceName
+ * @param {} deviceId
+ * @param {} activators
+ */
+function populateDeviceActivators(deviceName, deviceId, activators) {
```

Comment:
Really huge function, might want to split it up.

Reply:
👍 This is replaced in the new frontend.